SB-231 (feat) : QueryDSL 중고 게시글 검색 기능 개발

seob7 opened the pull request:
# SB-231 (feat) : QueryDSL 중고 게시글 검색 기능 개발

## 자신이 작성한 중고 판매 글 조회

- URL : `http://localhost/api/usedgoods`

- Method : GET

## 전체 중고 판매 글 조회 

- URL : `http://localhost/open-api/usedgoods`

- Method : GET

## 정렬 조건

- Default : size = 10, sort = postedAt,DESC

- `http://localhost/open-api/usedgoods/size=10&sort=postedAt,DESC`
    - 10개의 게시글을 조회하되 게시날짜를 기준으로 내림차순
- `http://localhost/open-api/usedgoods/search?size=3&sort=price,ASC`
    - 3개의 게시글을 조회하되 가격을 기준으로 오름차순

## 검색 조건

- keyword ( 제목 기준입니다. )

- minPrice

- maxPrice

- startDate

- endDate

### 1. 가격 범위로 조회

Parameter : minPrice, maxPrice

- `http://localhost/api/usedgoods?minPrice=1000`

    - 1000 이상 모두 조회

- `http://localhost/api/usedgoods?maxPrice=10000`

    - 10000 이상 모두 조회

- `http://localhost/api/usedgoods?minPrice=1000&maxPrice=10000`

    - 1000 이상 10000 이하 결과 조회

### 2. 날짜로 조회

Parameter : startDate, endDate

- `http://localhost/api/usedgoods?startDate=2024-08-01T00:00`

    - 2024/08/01 이후로 작성된 게시글 조회

- `http://localhost/api/usedgoods?endDate=2024-08-01T00:00`

    - 2024/08/01 이전에 작성된 게시글 조회

### 3. 제목 키워드로 검색

Parameter : keyword

- `http://localhost/api/usedgoods?keyword=노트북`

    - 제목에 “노트북” 이 들어간 결과 조회

### 4. 통합 검색

- `http://localhost/api/usedgoods?keyword=노트북&minPrice=100000&maxPrice=1500000&startDate=2024-08-01T00:00&endDate=2024-08-31T23:59`

    - 제목에 “노트북” 이 포함되고, 가격이 100000 이상 1500000 이하이며, 2024/08/01 부터 2024/08/31 까지 작성된 글을 조회함

### 5. 그 외

- 키워드와 가격 범위를 조합한 검색

- 날짜와 키워드를 조합한 검색

- 날짜와 가격 범위를 조합한 검색 

- 최소 가격과 날짜를 사용한 검색 

등 다양한 검색 조건을 만들 수 있습니다.

seob7 committed:
SB-232 (feat) : QueryDSL 환경 설정

SB-232 (feat) : QueryDSL 환경 설정

1. db/build.gradle
    - 빌드시 기존에 생성된 queryDir 제거
2 warehouse/build.gradle
3. QuerydslConfig

seob7 committed:
SB-233 (feat) : QueryUsedGoodsRepository 개발

SB-233 (feat) : QueryUsedGoodsRepository 개발

1. QueryUsedGoodsRepository 구현
2. EntitySearchCondition 구현

seob7 committed:
SB-234 (feat) : 중고 게시글 검색 기능 개발

SB-234 (feat) : 중고 게시글 검색 기능 개발

자신이 작성한 글, 전체 글 조회 기능 구현
1. Controller
    - UsedGoodsApiController
    - UsedGoodsOpenApiController
2. UsedGoodsBusiness
3. UsedGoodsConverter
4. UsedGoodsService
5. SearchCondition
6. UsedGoodsSearchResponse

seob7 committed:
SB-234 (feat) : 조회 Exception 처리

SB-234 (feat) : 조회 Exception 처리

1. UsedGoodsService
    - 검색결과가 존재하지 않을 때 Exception 발생

shinywoon commented on warehouse/src/main/java/warehouse/domain/usedgoods/business/UsedGoodsBusiness.java:
```diff
@@ -130,6 +136,29 @@ public MessageResponse buyUsedGoods(Long usedGoodsId, String email) {
+    public List<UsedGoodsSearchResponse> usedGoodsSearchBy(SearchCondition condition, Pageable page,
+        String email) {
+
+        Long userId = email != null ? usersService.getUserWithThrow(email).getId() : null;
```
`usersService.getUserWithThrow(email)` 이 메소드에서 null Check를 하고 있습니다.
수정이 필요합니다.

shinywoon commented on warehouse/src/main/java/warehouse/domain/usedgoods/controller/UsedGoodsApiController.java:
```diff
@@ -63,4 +70,16 @@ public Api<MessageResponse> buyUsedGoods(@AuthenticationPrincipal User user, @Pa
         return Api.OK(response);
     }
 
+    @GetMapping("/search")
```
`/search` Operation 은 불필요한 것 같습니다.
'GET /api/usedgoods` 로 심플하게 처리하면 됩니다.

seob7 committed:
SB-234 (refactor) : 물품 검색 API 변경

SB-234 (refactor) : 물품 검색 API 변경

GET /api/usedgoods/search -> GET /api/usedgoods 로 변경
1. UsedGoodsApiController
2. UsedGoodsOpenApiController

seob7 committed:
SB-234 (refactor) : Business & Converter 리팩토링

SB-234 (refactor) : Business & Converter 리팩토링

1. QueryUsedGoodsRepository
- 불필요한 Log 제거
2. UsedGoodsBusiness
- GoodsEntity 를 Map 으로 변환하여 각 UsedGoodsEntity 에 대응하는 GoodsEntity 를 조회하도록 변경
- return 에 스트림 .toList() 를 사용하여 바로 반환하도록 변경
3. UsedGoodsConverter
- 기존에 사용하던 toResponse 메서드 제거
- toSearchResponse 를 public 으로 변경하여 사용하도록 함
